// ngrx-data-pagination/esm2015/lib/pagination/store-interfaces/ngrx/pagination.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EntityOp } from '@ngrx/data';
import { select } from '@ngrx/store';
import { from } from 'rxjs';
import { filter, take } from 'rxjs/operators';
import { observableToPromisePaginationFunction, } from '../../iterator/pagination-function';
import { StorePaginationContext } from '../store-pagination-context';
import { allPaginationObservables, allPaginationSelectors, } from './selectors';
/**
 * Adapts `StorePaginationContext` to work with ngrx/data
 * @template Entity, NextPageState
 */
export class Pagination {
    /**
     * @param {?} contextId
     * @param {?} paginationFunction
     * @param {?} entityService
     * @param {?} store
     * @param {?} ngrxDataPaginationStoreKey
     * @param {?} addToCache
     */
    constructor(contextId, paginationFunction, entityService, store, ngrxDataPaginationStoreKey, addToCache) {
        this.entityService = entityService;
        /** @type {?} */
        const dispatch = (/**
         * @param {?} action
         * @return {?}
         */
        (action) => store.dispatch(action));
        /** @type {?} */
        const onReceivePage = (/**
         * @param {?} page
         * @return {?}
         */
        (page) => {
            if (!addToCache) {
                return;
            }
            this.entityService.upsertManyInCache(page);
            this.entityService.createAndDispatch(EntityOp.SAVE_UPSERT_MANY_SUCCESS, page);
        });
        /** @type {?} */
        const paginationState$ = (/** @type {?} */ (store.pipe(select(ngrxDataPaginationStoreKey))));
        /** @type {?} */
        const entityMap$ = entityService.selectors$.entityMap$;
        this.storePaginationContext = new StorePaginationContext(contextId, observableToPromisePaginationFunction(paginationFunction), dispatch, onReceivePage, paginationState$, entityMap$);
        this.selectors = allPaginationSelectors(contextId, entityService.selectors.selectEntityMap);
        this.selectors$ = allPaginationObservables(store, this.selectors);
    }
    /**
     * @return {?}
     */
    nextPage() {
        return this.storePaginationContext.nextPage();
    }
    /**
     * @return {?}
     */
    prevPage() {
        return this.storePaginationContext.prevPage();
    }
    /**
     * Not yet cancellable! Use with caution.
     * @return {?}
     */
    _loadAllPages() {
        return from(this._loadAllPagesP()).pipe(take(1));
    }
    /**
     * @private
     * @return {?}
     */
    _loadAllPagesP() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // wait until not loading
            yield this.selectors$.nextPageLoading
                .pipe(filter((/**
             * @param {?} loading
             * @return {?}
             */
            loading => !loading)), take(1))
                .toPromise();
            // start the next-page request
            this.nextPage();
            // wait until not loading, again
            yield this.selectors$.nextPageLoading
                .pipe(filter((/**
             * @param {?} loading
             * @return {?}
             */
            loading => !loading)), take(1))
                .toPromise();
            /** @type {?} */
            const done = yield this.selectors$.done
                .pipe(take(1))
                .toPromise();
            if (done) {
                // done case: let the whole function return
                return;
            }
            // recursive case: not done, so recurse
            return this._loadAllPages();
        });
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Pagination.prototype.storePaginationContext;
    /** @type {?} */
    Pagination.prototype.selectors;
    /** @type {?} */
    Pagination.prototype.selectors$;
    /**
     * @type {?}
     * @private
     */
    Pagination.prototype.entityService;
}
