// ngrx-data-pagination/esm2015/lib/pagination/store-interfaces/ngrx/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, select, } from '@ngrx/store';
import { contextSelectors } from '../../store/selectors';
import { defaultStoreKey } from './default-store-key';
/**
 * @record
 * @template Entity
 */
function AdvancedSelections() { }
if (false) {
    /** @type {?} */
    AdvancedSelections.prototype.page;
    /** @type {?} */
    AdvancedSelections.prototype.all;
}
/** @type {?} */
const mapValues = (/**
 * @template A
 * @param {?} obj
 * @param {?} f
 * @return {?}
 */
(obj, f) => {
    /** @type {?} */
    const result = {};
    for (const k of Object.keys(obj)) {
        result[k] = f(obj[k]);
    }
    return (/** @type {?} */ (result));
});
const ɵ0 = mapValues;
/**
 * Creates pagination selectors that only know about the pagination state
 * @type {?}
 */
const basicPaginationSelectors = (/**
 * @param {?} contextId
 * @return {?}
 */
(contextId) => {
    /** @type {?} */
    const paginationState = createFeatureSelector(defaultStoreKey);
    /** @type {?} */
    const contextState = createSelector(paginationState, (/**
     * @param {?} s
     * @return {?}
     */
    s => s.contexts[contextId]));
    /** @type {?} */
    const contextSelector = (/**
     * @template A
     * @param {?} f
     * @return {?}
     */
    (f) => createSelector(contextState, f));
    return mapValues(contextSelectors, contextSelector);
});
const ɵ1 = basicPaginationSelectors;
/**
 * Creates selectors that know about the entity cache
 * @type {?}
 */
const advancedPaginationSelectors = (/**
 * @template Entity
 * @param {?} basicPaginationSelectors
 * @param {?} selectEntityMap
 * @return {?}
 */
(basicPaginationSelectors, selectEntityMap) => {
    return {
        page: createSelector(basicPaginationSelectors.currentPageIds, selectEntityMap, (/**
         * @param {?} ids
         * @param {?} entityMap
         * @return {?}
         */
        (ids, entityMap) => {
            if (!ids || !entityMap) {
                return null;
            }
            return ids.reduce((/**
             * @param {?} a
             * @param {?} id
             * @return {?}
             */
            (a, id) => {
                if (id in entityMap) {
                    a.push(entityMap[id]);
                }
                return a;
            }), []);
        })),
        all: createSelector(basicPaginationSelectors.allIds, selectEntityMap, (/**
         * @param {?} allIds
         * @param {?} entityMap
         * @return {?}
         */
        (allIds, entityMap) => {
            if (!allIds || !entityMap) {
                return null;
            }
            return allIds.map((/**
             * @param {?} id
             * @return {?}
             */
            id => entityMap[id]));
        })),
    };
});
const ɵ2 = advancedPaginationSelectors;
/** @type {?} */
export const allPaginationSelectors = (/**
 * @template Entity
 * @param {?} contextId
 * @param {?} selectEntityMap
 * @return {?}
 */
(contextId, selectEntityMap) => {
    /** @type {?} */
    const basicSelectors = basicPaginationSelectors(contextId);
    /** @type {?} */
    const advancedSelectors = advancedPaginationSelectors(basicSelectors, selectEntityMap);
    return Object.assign({}, basicSelectors, advancedSelectors);
});
/** @type {?} */
export const allPaginationObservables = (/**
 * @template Entity
 * @param {?} store
 * @param {?} allSelectors
 * @return {?}
 */
(store, allSelectors) => {
    /** @type {?} */
    const sel = (/**
     * @template A
     * @param {?} selector
     * @return {?}
     */
    (selector) => store.pipe(select(selector)));
    return mapValues(allSelectors, sel);
});
export { ɵ0, ɵ1, ɵ2 };
