// ngrx-data-pagination/esm5/lib/pagination/iterator/async-page-generator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// TODO fix TS setup for AsyncPageGenerator types
// export type AsyncPageGenerator<Entity> = AsyncGenerator<Entity[], Entity[], any>;
/**
 * @template Entity, State
 * @param {?} paginationFunction
 * @return {?}
 */
export function asyncPageGenerator(paginationFunction) {
    return tslib_1.__asyncGenerator(this, arguments, function asyncPageGenerator_1() {
        var page;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // : AsyncGenerator<Entity[], Entity[], any>
                    page = {
                        data: undefined,
                        state: undefined,
                        done: false,
                    };
                    _a.label = 1;
                case 1:
                    if (!true) return [3 /*break*/, 7];
                    return [4 /*yield*/, tslib_1.__await(paginationFunction(page.state))];
                case 2:
                    page = _a.sent();
                    if (!page.done) return [3 /*break*/, 4];
                    return [4 /*yield*/, tslib_1.__await(page.data)];
                case 3: return [2 /*return*/, _a.sent()];
                case 4: return [4 /*yield*/, tslib_1.__await(page.data)];
                case 5: return [4 /*yield*/, _a.sent()];
                case 6:
                    _a.sent();
                    return [3 /*break*/, 1];
                case 7: return [2 /*return*/];
            }
        });
    });
}
