// ngrx-data-pagination/esm2015/lib/pagination/store/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const PaginationActionType = {
    RESET_PAGINATION_STATE: '[mb-Pagination] Reset Pagination State',
    GET_NEXT_PAGE: '[mb-Pagination] Get Next Page',
    GET_NEXT_PAGE_SUCCESS: '[mb-Pagination] Get Next Page Success',
    GET_NEXT_PAGE_EMPTY: '[mb-Pagination] Get Next Page Empty',
    PREV_PAGE: '[mb-Pagination] Prev Page',
    NEXT_PAGE: '[mb-Pagination] Next Page',
};
export { PaginationActionType };
/** @type {?} */
const T = PaginationActionType;
/**
 * @record
 */
function Action() { }
if (false) {
    /** @type {?} */
    Action.prototype.type;
}
/**
 * @record
 */
export function PaginationActionT() { }
if (false) {
    /** @type {?} */
    PaginationActionT.prototype.type;
    /** @type {?} */
    PaginationActionT.prototype.contextId;
}
export class ResetPaginationState {
    /**
     * @param {?} contextId
     */
    constructor(contextId) {
        this.contextId = contextId;
        this.type = T.RESET_PAGINATION_STATE;
    }
}
if (false) {
    /** @type {?} */
    ResetPaginationState.prototype.type;
    /** @type {?} */
    ResetPaginationState.prototype.contextId;
}
export class GetNextPage {
    /**
     * @param {?} contextId
     */
    constructor(contextId) {
        this.contextId = contextId;
        this.type = T.GET_NEXT_PAGE;
    }
}
if (false) {
    /** @type {?} */
    GetNextPage.prototype.type;
    /** @type {?} */
    GetNextPage.prototype.contextId;
}
export class GetNextPageSuccess {
    /**
     * @param {?} contextId
     * @param {?} entityIds
     * @param {?} done
     */
    constructor(contextId, entityIds, done) {
        this.contextId = contextId;
        this.entityIds = entityIds;
        this.done = done;
        this.type = T.GET_NEXT_PAGE_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    GetNextPageSuccess.prototype.type;
    /** @type {?} */
    GetNextPageSuccess.prototype.contextId;
    /** @type {?} */
    GetNextPageSuccess.prototype.entityIds;
    /** @type {?} */
    GetNextPageSuccess.prototype.done;
}
export class GetNextPageEmpty {
    /**
     * @param {?} contextId
     */
    constructor(contextId) {
        this.contextId = contextId;
        this.type = T.GET_NEXT_PAGE_EMPTY;
    }
}
if (false) {
    /** @type {?} */
    GetNextPageEmpty.prototype.type;
    /** @type {?} */
    GetNextPageEmpty.prototype.contextId;
}
export class PrevPage {
    /**
     * @param {?} contextId
     */
    constructor(contextId) {
        this.contextId = contextId;
        this.type = T.PREV_PAGE;
    }
}
if (false) {
    /** @type {?} */
    PrevPage.prototype.type;
    /** @type {?} */
    PrevPage.prototype.contextId;
}
export class NextPage {
    /**
     * @param {?} contextId
     */
    constructor(contextId) {
        this.contextId = contextId;
        this.type = T.NEXT_PAGE;
    }
}
if (false) {
    /** @type {?} */
    NextPage.prototype.type;
    /** @type {?} */
    NextPage.prototype.contextId;
}
/** @type {?} */
export const makeActionCreators = (/**
 * @param {?} contextId
 * @return {?}
 */
(contextId) => ({
    ResetPaginationState: (/**
     * @return {?}
     */
    () => new ResetPaginationState(contextId)),
    GetNextPage: (/**
     * @return {?}
     */
    () => new GetNextPage(contextId)),
    GetNextPageSuccess: (/**
     * @param {?} entityIds
     * @param {?} done
     * @return {?}
     */
    (entityIds, done) => new GetNextPageSuccess(contextId, entityIds, done)),
    GetNextPageEmpty: (/**
     * @return {?}
     */
    () => new GetNextPageEmpty(contextId)),
    PrevPage: (/**
     * @return {?}
     */
    () => new PrevPage(contextId)),
    NextPage: (/**
     * @return {?}
     */
    () => new NextPage(contextId)),
}));
/** @type {?} */
export const makeDispatchers = (/**
 * @param {?} contextId
 * @param {?} dispatch
 * @return {?}
 */
(contextId, dispatch) => {
    /** @type {?} */
    const actionCreators = makeActionCreators(contextId);
    return {
        ResetPaginationState: (/**
         * @return {?}
         */
        () => dispatch(actionCreators.ResetPaginationState())),
        GetNextPage: (/**
         * @return {?}
         */
        () => dispatch(actionCreators.GetNextPage())),
        GetNextPageSuccess: (/**
         * @param {?} entityIds
         * @param {?} done
         * @return {?}
         */
        (entityIds, done) => dispatch(actionCreators.GetNextPageSuccess(entityIds, done))),
        GetNextPageEmpty: (/**
         * @return {?}
         */
        () => dispatch(actionCreators.GetNextPageEmpty())),
        PrevPage: (/**
         * @return {?}
         */
        () => dispatch(actionCreators.PrevPage())),
        NextPage: (/**
         * @return {?}
         */
        () => dispatch(actionCreators.NextPage())),
    };
});
