// ngrx-data-pagination/esm2015/lib/pagination/store/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const finalPageIndex = (/**
 * @param {?} __0
 * @return {?}
 */
({ done, pages }) => {
    if (!done || !pages || !pages.length) {
        return null;
    }
    return pages.length - 1;
});
const ɵ0 = finalPageIndex;
/** @type {?} */
const flatten = (/**
 * @template A
 * @param {?} arrays
 * @return {?}
 */
(arrays) => arrays.reduce((/**
 * @param {?} result
 * @param {?} as
 * @return {?}
 */
(result, as) => result.concat(as))));
const ɵ1 = flatten;
const ɵ2 = /**
 * @param {?} __0
 * @return {?}
 */
({ pages, currentPage, }) => (pages ? pages[currentPage] : null), ɵ3 = /**
 * @param {?} __0
 * @return {?}
 */
({ pages }) => flatten(pages), ɵ4 = /**
 * @param {?} __0
 * @return {?}
 */
({ pages }) => pages, ɵ5 = /**
 * @param {?} __0
 * @return {?}
 */
({ pages, currentPage }) => !!pages[currentPage + 1], ɵ6 = /**
 * @param {?} __0
 * @return {?}
 */
({ pages, currentPage, loadingNewPage, }) => {
    if (!Number.isInteger(currentPage)) {
        return loadingNewPage;
    }
    return !pages[currentPage + 1] && loadingNewPage;
}, ɵ7 = /**
 * @param {?} __0
 * @return {?}
 */
({ loadingNewPage }) => loadingNewPage, ɵ8 = /**
 * @param {?} __0
 * @return {?}
 */
({ currentPage }) => currentPage >= 0 ? currentPage : null, ɵ9 = /**
 * @param {?} __0
 * @return {?}
 */
({ done }) => done, ɵ10 = /**
 * @param {?} __0
 * @return {?}
 */
({ currentPage }) => currentPage === 0, ɵ11 = /**
 * @param {?} state
 * @return {?}
 */
(state) => {
    const { currentPage } = state;
    /** @type {?} */
    const finalIndex = finalPageIndex(state);
    if (typeof finalIndex !== 'number' || typeof currentPage !== 'number') {
        return false;
    }
    return currentPage === finalIndex;
};
/** @type {?} */
export const contextSelectors = {
    currentPageIds: (ɵ2),
    allIds: (ɵ3),
    allPageIds: (ɵ4),
    nextPageLoaded: (ɵ5),
    nextPageLoading: (ɵ6),
    loadingNewPage: (ɵ7),
    pageNumber: (ɵ8),
    done: (ɵ9),
    finalPageIndex,
    onFirstPage: (ɵ10),
    onFinalPage: (ɵ11),
};
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11 };
