// ngrx-data-pagination/esm5/lib/pagination/store/actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var PaginationActionType = {
    RESET_PAGINATION_STATE: '[mb-Pagination] Reset Pagination State',
    GET_NEXT_PAGE: '[mb-Pagination] Get Next Page',
    GET_NEXT_PAGE_SUCCESS: '[mb-Pagination] Get Next Page Success',
    GET_NEXT_PAGE_EMPTY: '[mb-Pagination] Get Next Page Empty',
    PREV_PAGE: '[mb-Pagination] Prev Page',
    NEXT_PAGE: '[mb-Pagination] Next Page',
};
export { PaginationActionType };
/** @type {?} */
var T = PaginationActionType;
/**
 * @record
 */
function Action() { }
if (false) {
    /** @type {?} */
    Action.prototype.type;
}
/**
 * @record
 */
export function PaginationActionT() { }
if (false) {
    /** @type {?} */
    PaginationActionT.prototype.type;
    /** @type {?} */
    PaginationActionT.prototype.contextId;
}
var ResetPaginationState = /** @class */ (function () {
    function ResetPaginationState(contextId) {
        this.contextId = contextId;
        this.type = T.RESET_PAGINATION_STATE;
    }
    return ResetPaginationState;
}());
export { ResetPaginationState };
if (false) {
    /** @type {?} */
    ResetPaginationState.prototype.type;
    /** @type {?} */
    ResetPaginationState.prototype.contextId;
}
var GetNextPage = /** @class */ (function () {
    function GetNextPage(contextId) {
        this.contextId = contextId;
        this.type = T.GET_NEXT_PAGE;
    }
    return GetNextPage;
}());
export { GetNextPage };
if (false) {
    /** @type {?} */
    GetNextPage.prototype.type;
    /** @type {?} */
    GetNextPage.prototype.contextId;
}
var GetNextPageSuccess = /** @class */ (function () {
    function GetNextPageSuccess(contextId, entityIds, done) {
        this.contextId = contextId;
        this.entityIds = entityIds;
        this.done = done;
        this.type = T.GET_NEXT_PAGE_SUCCESS;
    }
    return GetNextPageSuccess;
}());
export { GetNextPageSuccess };
if (false) {
    /** @type {?} */
    GetNextPageSuccess.prototype.type;
    /** @type {?} */
    GetNextPageSuccess.prototype.contextId;
    /** @type {?} */
    GetNextPageSuccess.prototype.entityIds;
    /** @type {?} */
    GetNextPageSuccess.prototype.done;
}
var GetNextPageEmpty = /** @class */ (function () {
    function GetNextPageEmpty(contextId) {
        this.contextId = contextId;
        this.type = T.GET_NEXT_PAGE_EMPTY;
    }
    return GetNextPageEmpty;
}());
export { GetNextPageEmpty };
if (false) {
    /** @type {?} */
    GetNextPageEmpty.prototype.type;
    /** @type {?} */
    GetNextPageEmpty.prototype.contextId;
}
var PrevPage = /** @class */ (function () {
    function PrevPage(contextId) {
        this.contextId = contextId;
        this.type = T.PREV_PAGE;
    }
    return PrevPage;
}());
export { PrevPage };
if (false) {
    /** @type {?} */
    PrevPage.prototype.type;
    /** @type {?} */
    PrevPage.prototype.contextId;
}
var NextPage = /** @class */ (function () {
    function NextPage(contextId) {
        this.contextId = contextId;
        this.type = T.NEXT_PAGE;
    }
    return NextPage;
}());
export { NextPage };
if (false) {
    /** @type {?} */
    NextPage.prototype.type;
    /** @type {?} */
    NextPage.prototype.contextId;
}
/** @type {?} */
export var makeActionCreators = (/**
 * @param {?} contextId
 * @return {?}
 */
function (contextId) { return ({
    ResetPaginationState: (/**
     * @return {?}
     */
    function () { return new ResetPaginationState(contextId); }),
    GetNextPage: (/**
     * @return {?}
     */
    function () { return new GetNextPage(contextId); }),
    GetNextPageSuccess: (/**
     * @param {?} entityIds
     * @param {?} done
     * @return {?}
     */
    function (entityIds, done) {
        return new GetNextPageSuccess(contextId, entityIds, done);
    }),
    GetNextPageEmpty: (/**
     * @return {?}
     */
    function () { return new GetNextPageEmpty(contextId); }),
    PrevPage: (/**
     * @return {?}
     */
    function () { return new PrevPage(contextId); }),
    NextPage: (/**
     * @return {?}
     */
    function () { return new NextPage(contextId); }),
}); });
/** @type {?} */
export var makeDispatchers = (/**
 * @param {?} contextId
 * @param {?} dispatch
 * @return {?}
 */
function (contextId, dispatch) {
    /** @type {?} */
    var actionCreators = makeActionCreators(contextId);
    return {
        ResetPaginationState: (/**
         * @return {?}
         */
        function () { return dispatch(actionCreators.ResetPaginationState()); }),
        GetNextPage: (/**
         * @return {?}
         */
        function () { return dispatch(actionCreators.GetNextPage()); }),
        GetNextPageSuccess: (/**
         * @param {?} entityIds
         * @param {?} done
         * @return {?}
         */
        function (entityIds, done) {
            return dispatch(actionCreators.GetNextPageSuccess(entityIds, done));
        }),
        GetNextPageEmpty: (/**
         * @return {?}
         */
        function () { return dispatch(actionCreators.GetNextPageEmpty()); }),
        PrevPage: (/**
         * @return {?}
         */
        function () { return dispatch(actionCreators.PrevPage()); }),
        NextPage: (/**
         * @return {?}
         */
        function () { return dispatch(actionCreators.NextPage()); }),
    };
});
