// ngrx-data-pagination/esm5/lib/pagination/store/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var finalPageIndex = (/**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var done = _a.done, pages = _a.pages;
    if (!done || !pages || !pages.length) {
        return null;
    }
    return pages.length - 1;
});
var ɵ0 = finalPageIndex;
/** @type {?} */
var flatten = (/**
 * @template A
 * @param {?} arrays
 * @return {?}
 */
function (arrays) { return arrays.reduce((/**
 * @param {?} result
 * @param {?} as
 * @return {?}
 */
function (result, as) { return result.concat(as); })); });
var ɵ1 = flatten;
var ɵ2 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var pages = _a.pages, currentPage = _a.currentPage;
    return (pages ? pages[currentPage] : null);
}, ɵ3 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var pages = _a.pages;
    return flatten(pages);
}, ɵ4 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var pages = _a.pages;
    return pages;
}, ɵ5 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var pages = _a.pages, currentPage = _a.currentPage;
    return !!pages[currentPage + 1];
}, ɵ6 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var pages = _a.pages, currentPage = _a.currentPage, loadingNewPage = _a.loadingNewPage;
    if (!Number.isInteger(currentPage)) {
        return loadingNewPage;
    }
    return !pages[currentPage + 1] && loadingNewPage;
}, ɵ7 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var loadingNewPage = _a.loadingNewPage;
    return loadingNewPage;
}, ɵ8 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var currentPage = _a.currentPage;
    return currentPage >= 0 ? currentPage : null;
}, ɵ9 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var done = _a.done;
    return done;
}, ɵ10 = /**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var currentPage = _a.currentPage;
    return currentPage === 0;
}, ɵ11 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    var currentPage = state.currentPage;
    /** @type {?} */
    var finalIndex = finalPageIndex(state);
    if (typeof finalIndex !== 'number' || typeof currentPage !== 'number') {
        return false;
    }
    return currentPage === finalIndex;
};
/** @type {?} */
export var contextSelectors = {
    currentPageIds: (ɵ2),
    allIds: (ɵ3),
    allPageIds: (ɵ4),
    nextPageLoaded: (ɵ5),
    nextPageLoading: (ɵ6),
    loadingNewPage: (ɵ7),
    pageNumber: (ɵ8),
    done: (ɵ9),
    finalPageIndex: finalPageIndex,
    onFirstPage: (ɵ10),
    onFinalPage: (ɵ11),
};
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11 };
