// ngrx-data-pagination/esm2015/lib/pagination/store-interfaces/store-pagination-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { combineLatest } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import { PageIterator } from '../iterator/page-iterator';
import { makeDispatchers } from '../store/actions';
import { contextSelectors } from '../store/selectors';
import { defaultPaginationContextState, } from '../store/state';
// assumes paginator reducer is plugged in
/**
 * This is a pagination context usable with Redux, ngrx, or
 * any other similar implementations. It writes to, but does
 * not read from the store. It assumes:
 * 1. `dispatch` dispatches an action to the store
 * 2. TODO this library's reducer has been installed
 * @template Entity
 */
export class StorePaginationContext {
    /**
     * @param {?} contextId
     * @param {?} paginationFunction
     * @param {?} dispatch
     * @param {?} onReceivePage
     * @param {?} state$
     * @param {?} entityMap$
     */
    constructor(
    // Arbitrary. For now, only use one ReduxLikePaginationContext per contextId
    contextId, 
    // For requesting the pages
    paginationFunction, 
    // Dispatch an action meant for the PaginationReducer
    dispatch, onReceivePage, state$, entityMap$) {
        this.onReceivePage = onReceivePage;
        this.contextState = defaultPaginationContextState;
        this[Symbol.asyncIterator] = (/**
         * @return {?}
         */
        () => this.pageIterator[Symbol.asyncIterator]);
        this.contextState$ = state$.pipe(map((/**
         * @param {?} state
         * @return {?}
         */
        state => state ? state.contexts[contextId] : defaultPaginationContextState)));
        this.subscription = this.contextState$.subscribe((/**
         * @param {?} contextState
         * @return {?}
         */
        contextState => {
            this.contextState = contextState;
        }));
        this.dispatchers = makeDispatchers(contextId, dispatch);
        this.dispatchers.ResetPaginationState();
        this.pageIterator = new PageIterator(paginationFunction);
        this.entityMap$ = entityMap$.pipe(shareReplay(1));
        // wait until next event loop in case of setup time
        setTimeout((/**
         * @return {?}
         */
        () => this.nextPage()), 0);
    }
    /**
     * @return {?}
     */
    getNextPageP() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const nextPageLoaded = contextSelectors.nextPageLoaded(this.contextState);
            if (nextPageLoaded) {
                this.incrementCurrentPage();
                return;
            }
            if (contextSelectors.done(this.contextState)) {
                throw new Error('Cannot get more pages after becoming done');
            }
            this.dispatchers.GetNextPage();
            /** @type {?} */
            const page = yield this.pageIterator.getNextPage();
            this.onReceivePage(page);
            if (!page) {
                throw new Error('bad page in getNextPageP');
            }
            /** @type {?} */
            const entityIds = page.map((/**
             * @param {?} __0
             * @return {?}
             */
            ({ id }) => id));
            if (!entityIds.length) {
                this.dispatchers.GetNextPageEmpty();
            }
            else {
                this.dispatchers.GetNextPageSuccess(entityIds, this.pageIterator.done);
            }
            return page;
        });
    }
    /**
     * @return {?}
     */
    nextPage() {
        this.getNextPageP();
    }
    /**
     * @return {?}
     */
    prevPage() {
        if (this.contextState.currentPage <= 0) {
            throw new Error('Cannot go back from page 0');
        }
        this.dispatchers.PrevPage();
    }
    /**
     * @private
     * @return {?}
     */
    incrementCurrentPage() {
        this.dispatchers.NextPage();
    }
    /**
     * @return {?}
     */
    get currentPage$() {
        return combineLatest(this.entityMap$, this.contextState$).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([entityMap, contextState]) => {
            /** @type {?} */
            const currentPageIds = contextSelectors.currentPageIds(contextState);
            if (!currentPageIds) {
                return null;
            }
            return currentPageIds.map((/**
             * @param {?} entityId
             * @return {?}
             */
            entityId => entityMap[entityId]));
        })));
    }
    /**
     * @return {?}
     */
    destroy() {
        this.subscription.unsubscribe();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.pageIterator;
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.dispatchers;
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.contextState;
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.contextState$;
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.entityMap$;
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.subscription;
    /* Skipping unnamed member:
    [Symbol.asyncIterator] = () => this.pageIterator[Symbol.asyncIterator];*/
    /**
     * @type {?}
     * @private
     */
    StorePaginationContext.prototype.onReceivePage;
}
