// ngrx-data-pagination/esm5/lib/pagination/store-interfaces/ngrx/pagination.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, NgModule, Optional, SkipSelf, } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { paginationReducer } from '../../store/reducer';
import { defaultStoreKey } from './default-store-key';
import { PaginationFactory } from './pagination-factory.service';
/** @type {?} */
var NGRX_DATA_PAGINATION_STORE_KEY = 'NGRX_DATA_PAGINATION_STORE_KEY';
var PaginationModule = /** @class */ (function () {
    function PaginationModule(parentModule, storeKey) {
        if (parentModule) {
            throw new Error('PaginationModule is already loaded. Import it in the root module (usually the AppModule) only');
        }
        if (!storeKey) {
            throw new Error('ngrx-data-pagination requires PaginationModule.forRoot() to be imported');
        }
    }
    /**
     * `storeKey` doesn't actually work yet`
     */
    /**
     * `storeKey` doesn't actually work yet`
     * @return {?}
     */
    PaginationModule.forRoot = /**
     * `storeKey` doesn't actually work yet`
     * @return {?}
     */
    function () {
        return {
            ngModule: PaginationModule,
            providers: [
                {
                    provide: NGRX_DATA_PAGINATION_STORE_KEY,
                    useValue: defaultStoreKey,
                },
            ],
        };
    };
    PaginationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [StoreModule.forFeature(defaultStoreKey, paginationReducer)],
                    exports: [],
                    providers: [PaginationFactory],
                },] }
    ];
    /** @nocollapse */
    PaginationModule.ctorParameters = function () { return [
        { type: PaginationModule, decorators: [{ type: Optional }, { type: SkipSelf }] },
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [NGRX_DATA_PAGINATION_STORE_KEY,] }] }
    ]; };
    return PaginationModule;
}());
export { PaginationModule };
