// ngrx-data-pagination/esm5/lib/pagination/store-interfaces/ngrx/pagination.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EntityOp } from '@ngrx/data';
import { select } from '@ngrx/store';
import { from } from 'rxjs';
import { filter, take } from 'rxjs/operators';
import { observableToPromisePaginationFunction, } from '../../iterator/pagination-function';
import { StorePaginationContext } from '../store-pagination-context';
import { allPaginationObservables, allPaginationSelectors, } from './selectors';
/**
 * Adapts `StorePaginationContext` to work with ngrx/data
 * @template Entity, NextPageState
 */
var /**
 * Adapts `StorePaginationContext` to work with ngrx/data
 * @template Entity, NextPageState
 */
Pagination = /** @class */ (function () {
    function Pagination(contextId, paginationFunction, entityService, store, ngrxDataPaginationStoreKey, addToCache) {
        var _this = this;
        this.entityService = entityService;
        /** @type {?} */
        var dispatch = (/**
         * @param {?} action
         * @return {?}
         */
        function (action) { return store.dispatch(action); });
        /** @type {?} */
        var onReceivePage = (/**
         * @param {?} page
         * @return {?}
         */
        function (page) {
            if (!addToCache) {
                return;
            }
            _this.entityService.upsertManyInCache(page);
            _this.entityService.createAndDispatch(EntityOp.SAVE_UPSERT_MANY_SUCCESS, page);
        });
        /** @type {?} */
        var paginationState$ = (/** @type {?} */ (store.pipe(select(ngrxDataPaginationStoreKey))));
        /** @type {?} */
        var entityMap$ = entityService.selectors$.entityMap$;
        this.storePaginationContext = new StorePaginationContext(contextId, observableToPromisePaginationFunction(paginationFunction), dispatch, onReceivePage, paginationState$, entityMap$);
        this.selectors = allPaginationSelectors(contextId, entityService.selectors.selectEntityMap);
        this.selectors$ = allPaginationObservables(store, this.selectors);
    }
    /**
     * @return {?}
     */
    Pagination.prototype.nextPage = /**
     * @return {?}
     */
    function () {
        return this.storePaginationContext.nextPage();
    };
    /**
     * @return {?}
     */
    Pagination.prototype.prevPage = /**
     * @return {?}
     */
    function () {
        return this.storePaginationContext.prevPage();
    };
    /** Not yet cancellable! Use with caution. */
    /**
     * Not yet cancellable! Use with caution.
     * @return {?}
     */
    Pagination.prototype._loadAllPages = /**
     * Not yet cancellable! Use with caution.
     * @return {?}
     */
    function () {
        return from(this._loadAllPagesP()).pipe(take(1));
    };
    /**
     * @private
     * @return {?}
     */
    Pagination.prototype._loadAllPagesP = /**
     * @private
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var done;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: 
                    // wait until not loading
                    return [4 /*yield*/, this.selectors$.nextPageLoading
                            .pipe(filter((/**
                         * @param {?} loading
                         * @return {?}
                         */
                        function (loading) { return !loading; })), take(1))
                            .toPromise()];
                    case 1:
                        // wait until not loading
                        _a.sent();
                        // start the next-page request
                        this.nextPage();
                        // wait until not loading, again
                        return [4 /*yield*/, this.selectors$.nextPageLoading
                                .pipe(filter((/**
                             * @param {?} loading
                             * @return {?}
                             */
                            function (loading) { return !loading; })), take(1))
                                .toPromise()];
                    case 2:
                        // wait until not loading, again
                        _a.sent();
                        return [4 /*yield*/, this.selectors$.done
                                .pipe(take(1))
                                .toPromise()];
                    case 3:
                        done = _a.sent();
                        if (done) {
                            // done case: let the whole function return
                            return [2 /*return*/];
                        }
                        // recursive case: not done, so recurse
                        return [2 /*return*/, this._loadAllPages()];
                }
            });
        });
    };
    return Pagination;
}());
/**
 * Adapts `StorePaginationContext` to work with ngrx/data
 * @template Entity, NextPageState
 */
export { Pagination };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Pagination.prototype.storePaginationContext;
    /** @type {?} */
    Pagination.prototype.selectors;
    /** @type {?} */
    Pagination.prototype.selectors$;
    /**
     * @type {?}
     * @private
     */
    Pagination.prototype.entityService;
}
