// ngrx-data-pagination/esm5/lib/pagination/store-interfaces/ngrx/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { createFeatureSelector, createSelector, select, } from '@ngrx/store';
import { contextSelectors } from '../../store/selectors';
import { defaultStoreKey } from './default-store-key';
/**
 * @record
 * @template Entity
 */
function AdvancedSelections() { }
if (false) {
    /** @type {?} */
    AdvancedSelections.prototype.page;
    /** @type {?} */
    AdvancedSelections.prototype.all;
}
/** @type {?} */
var mapValues = (/**
 * @template A
 * @param {?} obj
 * @param {?} f
 * @return {?}
 */
function (obj, f) {
    var e_1, _a;
    /** @type {?} */
    var result = {};
    try {
        for (var _b = tslib_1.__values(Object.keys(obj)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var k = _c.value;
            result[k] = f(obj[k]);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return (/** @type {?} */ (result));
});
var ɵ0 = mapValues;
/**
 * Creates pagination selectors that only know about the pagination state
 * @type {?}
 */
var basicPaginationSelectors = (/**
 * @param {?} contextId
 * @return {?}
 */
function (contextId) {
    /** @type {?} */
    var paginationState = createFeatureSelector(defaultStoreKey);
    /** @type {?} */
    var contextState = createSelector(paginationState, (/**
     * @param {?} s
     * @return {?}
     */
    function (s) { return s.contexts[contextId]; }));
    /** @type {?} */
    var contextSelector = (/**
     * @template A
     * @param {?} f
     * @return {?}
     */
    function (f) {
        return createSelector(contextState, f);
    });
    return mapValues(contextSelectors, contextSelector);
});
var ɵ1 = basicPaginationSelectors;
/**
 * Creates selectors that know about the entity cache
 * @type {?}
 */
var advancedPaginationSelectors = (/**
 * @template Entity
 * @param {?} basicPaginationSelectors
 * @param {?} selectEntityMap
 * @return {?}
 */
function (basicPaginationSelectors, selectEntityMap) {
    return {
        page: createSelector(basicPaginationSelectors.currentPageIds, selectEntityMap, (/**
         * @param {?} ids
         * @param {?} entityMap
         * @return {?}
         */
        function (ids, entityMap) {
            if (!ids || !entityMap) {
                return null;
            }
            return ids.reduce((/**
             * @param {?} a
             * @param {?} id
             * @return {?}
             */
            function (a, id) {
                if (id in entityMap) {
                    a.push(entityMap[id]);
                }
                return a;
            }), []);
        })),
        all: createSelector(basicPaginationSelectors.allIds, selectEntityMap, (/**
         * @param {?} allIds
         * @param {?} entityMap
         * @return {?}
         */
        function (allIds, entityMap) {
            if (!allIds || !entityMap) {
                return null;
            }
            return allIds.map((/**
             * @param {?} id
             * @return {?}
             */
            function (id) { return entityMap[id]; }));
        })),
    };
});
var ɵ2 = advancedPaginationSelectors;
/** @type {?} */
export var allPaginationSelectors = (/**
 * @template Entity
 * @param {?} contextId
 * @param {?} selectEntityMap
 * @return {?}
 */
function (contextId, selectEntityMap) {
    /** @type {?} */
    var basicSelectors = basicPaginationSelectors(contextId);
    /** @type {?} */
    var advancedSelectors = advancedPaginationSelectors(basicSelectors, selectEntityMap);
    return tslib_1.__assign({}, basicSelectors, advancedSelectors);
});
/** @type {?} */
export var allPaginationObservables = (/**
 * @template Entity
 * @param {?} store
 * @param {?} allSelectors
 * @return {?}
 */
function (store, allSelectors) {
    /** @type {?} */
    var sel = (/**
     * @template A
     * @param {?} selector
     * @return {?}
     */
    function (selector) { return store.pipe(select(selector)); });
    return mapValues(allSelectors, sel);
});
export { ɵ0, ɵ1, ɵ2 };
