// ngrx-data-pagination/esm2015/lib/pagination/iterator/async-page-generator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// TODO fix TS setup for AsyncPageGenerator types
// export type AsyncPageGenerator<Entity> = AsyncGenerator<Entity[], Entity[], any>;
/**
 * @template Entity, State
 * @param {?} paginationFunction
 * @return {?}
 */
export function asyncPageGenerator(paginationFunction) {
    return tslib_1.__asyncGenerator(this, arguments, function* asyncPageGenerator_1() {
        // : AsyncGenerator<Entity[], Entity[], any>
        /** @type {?} */
        let page = {
            data: undefined,
            state: undefined,
            done: false,
        };
        while (true) {
            page = yield tslib_1.__await(paginationFunction(page.state));
            if (page.done) {
                return yield tslib_1.__await(page.data);
            }
            yield yield tslib_1.__await(page.data);
        }
    });
}
